### PR TITLE
S✔ ◾ 🐛 Fix - Screenshot caption missing + duplicated screenshots in generated issues (#834)

### LINKED ISSUE #834: 🐛 Bug - Screenshot caption missing + duplicated screenshots in generated issues
Cc: @calumjs @ricksu978 @adamcogan
Hi YakShaver Desktop team

[🟥 Watch the video (1 min 12 sec)](https://www.youtube.com/watch?v=40q9vFNBmUE)

### Pain
Issues created by YakShaver Desktop are formatting screenshots incorrectly:
- The screenshot at the top of the issue is missing the required caption text (e.g. “Figure: …”) underneath the image.
- The caption text should be bold and visually prominent (currently missing / not bold).
- The same screenshot is being duplicated (appears at the top of the issue and again in the “### Screenshots” section), creating clutter and confusion.

### Reproduce Steps
1. Use YakShaver Desktop to shave an issue into GitHub
2. View the created issue on GitHub.
3. Observe the screenshot added in two places: at the top, in the screenshots section

### Expected Behavior
- The screenshot is shown once in the issue body.
- The screenshot has a bold caption underneath it (e.g. **Figure: ...**) describing what is shown.
- The “### Screenshots” section does not contain a duplicated copy of the same screenshot when it is already embedded earlier in the issue.

### Acceptance Criteria
- [ ] Screenshot embedded by YakShaver Desktop includes a bold caption underneath (e.g. **Figure: ...**) describing the image
- [ ] Caption text is visually prominent (matches the current/expected formatting shown in the “Screenshots section” example)
- [ ] The same screenshot is not inserted twice in the generated issue body
- [ ] Creating an issue with one screenshot results in exactly one rendered screenshot in the final GitHub issue

### Tasks
- [ ] Investigate how YakShaver Desktop generates screenshot markdown and captions
- [ ] Identify why the caption text is missing under the top screenshot
- [ ] Update screenshot rendering to include a bold “Figure:” caption under the image
- [ ] Identify why screenshots are duplicated between the top of the issue and the “### Screenshots” section
- [ ] Remove/avoid duplicate insertion of the same screenshot in generated issues
- [ ] Add/adjust tests to cover screenshot caption formatting
- [ ] Add/adjust tests to cover screenshot duplication prevention
- [ ] Verify the fix by creating a new issue via YakShaver Desktop

### More Information
Reporter notes:
- A screenshot manually added in the “### Screenshots” section shows the desired result (it includes a Figure text/caption underneath describing what is in the screenshot).

### Environment
 - Device: N/A
 - Browser: N/A
 - OS: N/A

### Screenshots
YakShaver issue shows screenshot without caption and with duplication

Thanks!

### PR DESCRIPTION
**Pull request by calumjs**
Fixes #834

## Symptoms
1. **Missing caption** — the screenshot embedded by YakShaver Desktop had no bold `**Figure: ...**` caption beneath it.
2. **Duplicated screenshot** — the same screenshot appeared twice: once near the top of the issue body and again in the `### Screenshots` section.

## Root cause
The issue body markdown is authored freely by the LLM following `SHARED_ISSUE_CREATION_RULES` rule 8 in `src/backend/constants/prompts.ts`. That rule:
- emitted `Screenshot description` with **no caption**, and
- gave **no single-placement instruction**, so the model embedded the screenshot inline *and* also filled the issue template's `### Screenshots` section with the same image.

There is no deterministic TS function that assembles the GitHub issue body today — the body is produced by the model and sent straight to the external GitHub MCP `create_issue` tool during the agent loop.

## Fix
- **Prompt** (`prompts.ts`): rewrote rule 8 to require a bold `**Figure: ...**` caption on the line beneath the embed, and to place the screenshot in exactly ONE location (the `### Screenshots` section if present, otherwise once near the top) — never duplicated.
- **Deterministic guard** (`src/backend/utils/screenshot-markdown.ts`): added a pure `normalizeIssueScreenshots(body)` that de-dupes repeated embeds of the same image URL (keeps the first, drops later ones plus any orphaned caption) and ensures every remaining embed is followed by a bold `**Figure: ...**` caption (synthesised from the alt text, with a sensible fallback for generic/empty alts). This holds the invariants regardless of model phrasing.
- **Wiring** (`mcp-orchestrator.ts`): the guard runs in the pre-execute hook (alongside `resolveToolOutputReferences`) for backlog create/update tools, normalizing body-like fields (`body`/`description`/`content`/`markdown`/`text`) before the tool executes. Non-backlog tools are untouched. URL query params are preserved.

## Test
`src/backend/utils/screenshot-markdown.test.ts` — 10 deterministic tests covering both bugs: caption added when missing, duplicate embed removed (top + `### Screenshots`), orphaned caption cleanup, both bugs in one pass, distinct screenshots preserved, no double-caption, query-param preservation, and no-op on bodies without images.

## Verify (this branch)
- `npm run build` (tsc): green.
- `npx vitest run src/backend/utils/screenshot-markdown.test.ts`: 10/10 pass.
- Existing `mcp-orchestrator.false-success.test.ts`: 13/13 still pass.
- `npx @biomejs/biome@2.3.11 check` on changed files: clean.

## Pending live verification
**Live repro / walkthrough is pending** (a video fork owns the running app, so this was code + build + unit-test only). Needs a real shave-to-GitHub run to confirm the rendered issue shows exactly one captioned screenshot end-to-end.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by calumjs — COMMENTED** [review]
## Review summary

Three confirmed findings on the screenshot-caption-dedupe change, two high and one medium. They are interrelated: a brittle name-based gate, a dedup tiebreak that contradicts the steering prompt, and no test coverage over the new gating branch.

---

### High

**1. Backlog-tool gating is a brittle name regex — re-introduces the pattern #833 removed**
`src/backend/services/mcp/mcp-orchestrator.ts:171-183` (`normalizeScreenshotMarkdownInArgs` / `isBacklogMutation`)

The decision "is this a backlog create/update tool" is made by two hand-rolled keyword regexes over third-party MCP tool names: `/(issue|work[_-]?item|backlog|pbi|ticket|story|task)/` AND `/(create|update|add|new|edit|file|append)/`. This is the exact pattern this file deliberately moved away from in #833, where the old `BACKLOG_MUTATION_TOOL` regex was replaced by the LLM-as-judge `judgeBacklogOutcome` (still present in this file). Probed misfires on the codebase's own test tool names:

- `github__add_issue_comment` -> `true` (matches `issue` + `add`): the guard rewrites issue-COMMENT bodies, injecting `**Figure:**` captions and stripping repeated images from user-facing comments the PR's stated scope never intended to touch.
- `github__update_issue_cache` -> `true`: false-positive on a read-only cache tool (the existing false-success.test.ts already flags this shape as one "the old regex false-matched on").
- `jira__transition_issue` -> `false`: a real mutation tool that uses none of the create/update stems is silently skipped.

Both failure modes are silent. Recommend not deriving tool category from names: gate on a structured signal the system already produces (the MCP tool's mutating/read-only annotation, or the specific resolved tool ids the orchestrator registered). At minimum, replace open keyword spotting with an anchored allow-list of the actual create/update tools and explicitly exclude `*_comment`.

**2. Dedup keep-first contradicts the new prompt's placement intent**
`src/backend/utils/screenshot-markdown.ts:60-75` (`normalizeIssueScreenshots`) vs `prompts.ts` rule 8

Rule 8 (new in this PR) tells the model to put the single captioned screenshot in the `### Screenshots` section, which per `.github/ISSUE_TEMPLATE/1-bug.md` (line 48 of ~52) is near the bottom. But `normalizeIssueScreenshots` keeps the FIRST occurrence. Reproduced: a body with a stray `stray` at top plus the intended `### Screenshots\ngood shot\n**Figure: good shot**` at bottom normalizes to keep the top stray (synthesizing `**Figure: stray**`) and empties the `### Screenshots` section, leaving a dangling heading with no image — the opposite of what the prompt asks. Worse, when the top stray has a generic alt (`Screenshot`), the guard keeps `**Figure: Screenshot from the recording**` and discards a descriptive `**Figure: Login page crashes on submit**` from the Screenshots section, keeping the weaker caption. The PR's own test asserts only `toContain("### Screenshots")` (the bare heading), so the empty section is knowingly accepted. Make the dedup keep (or relocate the survivor into) the occurrence inside `### Screenshots` when one exists.

---

### Medium

**3. New gating branch is untested — only the pure transform is covered**
`src/backend/utils/screenshot-markdown.test.ts` (all 10 tests) + `mcp-orchestrator.ts` wiring

Every test exercises only the pure transforms `normalizeIssueScreenshots` / `buildFigureCaption`. There is zero coverage of `normalizeScreenshotMarkdownInArgs` or the `isBacklogMutation` gate — the actual new branch this PR adds. `git grep` for those symbols returns only the definition site. The sibling `mcp-orchestrator.false-success.test.ts` is unrelated (it drives `github__create_issue` with empty input and asserts only on `backlogActionSucceeded`/`execute`). Given finding #1 already misfires on a real tool name, this branch needs both-sides tests: assert it fires for `create_issue`/`update_issue` body fields, is a no-op for non-backlog tools (e.g. `search_repositories`, `get_file_contents`), and does NOT rewrite `add_issue_comment`.

File: src/backend/services/mcp/mcp-orchestrator.ts [hunk 1]
```diff
@@ -159,6 +160,45 @@ export class MCPOrchestrator {
...
+   * `**Figure: ...**` caption. We normalise the body-bearing string fields here so the rendered
+   * issue always has exactly one captioned screenshot, regardless of model phrasing.
+   */
+  private normalizeScreenshotMarkdownInArgs(
+    toolName: string,
+    input: Record<string, unknown>,
+  ): Record<string, unknown> {
+    // Only touch tools that create/update a backlog item (issue / work item / PBI / ticket).
+    const lowerName = toolName.toLowerCase();
+    const isBacklogMutation =
```
Comment (calumjs) [inline src/backend/services/mcp/mcp-orchestrator.ts]: `isBacklogMutation` decides tool category by two keyword regexes over third-party tool names. Probed misfires: `github__add_issue_comment` -> true (rewrites user-facing comment bodies), `github__update_issue_cache` -> true (read-only tool), `jira__transition_issue` -> false (real mutation skipped). This re-introduces the name-regex pattern #833 replaced with `judgeBacklogOutcome` (still in this file). Gate on a structured mutating/read-only signal or an anchored allow-list that excludes `*_comment`.

File: src/backend/utils/screenshot-markdown.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    if (!match?.groups) {
+      result.push(line);
+      continue;
+    }
+
+    const url = match.groups.url;
+
+    // Duplicate embed of an already-seen image — drop this line entirely.
+    if (seenUrls.has(url)) {
```
Comment (calumjs) [inline src/backend/utils/screenshot-markdown.ts]: Keep-first tiebreak contradicts new prompts.ts rule 8, which steers the single captioned screenshot into the `### Screenshots` section (bottom of the bug template, line ~48). When the model also leaves a stray top embed, this keeps the top stray and empties `### Screenshots`, leaving a dangling heading — and if the stray has a generic alt it keeps `**Figure: Screenshot from the recording**` while discarding the descriptive caption from the Screenshots section. Prefer keeping (or relocating the survivor into) the occurrence inside `### Screenshots` when one exists.

File: src/backend/services/mcp/mcp-orchestrator.ts (hunk 1 above)
Comment (calumjs) [reply src/backend/services/mcp/mcp-orchestrator.ts]: Fixed in 4bc5583. Replaced the two keyword regexes with an anchored allow-list `isBacklogItemMutationTool` (now a pure, exported helper in `screenshot-markdown.ts`):

- It excludes `comment|reply|reaction|cache|search|list|get|read|find|query|delete|close` first, so `add_issue_comment` and `update_issue_cache` no longer match.
- It then matches only names that END in a backlog noun (`issue`, `work_item`, `pbi`, `ticket`, `story`, `task`) AND contain a create/update verb. Anchoring the noun at the end is the discriminator: `add_issue_comment` ends in `comment` and `update_issue_cache` ends in `cache` (both excluded), while `create_or_update_file` / `create_pull_request` end in `file` / `request` (not backlog nouns).

On `jira__transition_issue`: it ends in `issue` but carries no create/update verb, so it stays out — consistent with this guard only rewriting bodies that author an issue/work-item, not state-transition tools (which don't carry a body to normalise).

Added unit tests asserting the both-sides behaviour (fires for create/update issue/work-item across GitHub/ADO/Jira; no-op for comment/cache/search/get) plus orchestrator wiring tests proving `normalizeScreenshotMarkdownInArgs` rewrites create-issue/update-work-item bodies and never touches `add_issue_comment` / `update_issue_cache`.

File: src/backend/utils/screenshot-markdown.ts (hunk 2 above)
Comment (calumjs) [reply src/backend/utils/screenshot-markdown.ts]: Fixed in 4bc5583. `normalizeIssueScreenshots` now picks the surviving occurrence of each duplicated URL via `chooseSurvivingOccurrences`: it prefers the embed inside the `### Screenshots` section when one exists, and falls back to keep-first otherwise. So when the model leaves a stray top embed plus the intended `### Screenshots` copy, the Screenshots copy survives — the section is no longer emptied, no dangling heading, and the descriptive caption there is kept instead of the weak `**Figure: Screenshot from the recording**` synthesised from the stray's alt. Added two tests covering exactly this (stray-top + Screenshots-section dedupe, including the generic-alt-vs-descriptive-caption case).

**Review by calumjs — COMMENTED** [review]
## Review: #834 screenshot caption + dedupe backstop

The deterministic guard and the dedupe/caption logic are reasonable, but the verification is one-sided and the gating predicate misses the tool name the configured GitHub MCP server actually emits. Net effect: on the live config the backstop is inert, and the connected happy-path outcome is unproven. Details below, grouped by severity.

### High

**1. The gating predicate never matches the live GitHub tool name (`issue_write`) — the backstop is dead code on the configured server.**
`src/backend/utils/screenshot-markdown.ts` — `isBacklogItemMutationTool` anchors the backlog noun at the END of the bare name:
```
const nounAtEnd = /(work[_-]?item|backlog[_-]?item|issue|pbi|ticket|story|task)$/;
```
The configured GitHub preset is `https://api.githubcopilot.com/mcp/` (`src/shared/mcp/preset-servers.ts:17`), i.e. the official github-mcp-server, which consolidated `create_issue`/`update_issue` into a single action-based `issue_write` tool (plus `sub_issue_write`). `GitHub__issue_write` strips to `issue_write`, which ends in `write`, not `issue`, so `nounAtEnd` fails and the function returns `false`. `normalizeScreenshotMarkdownInArgs` then early-returns and the dedupe/caption normalization is silently skipped for the real GitHub path. The repo already treats `issue_write` as canonical (`src/ui/src/utils/index.ts:136`). The unit tests only assert the legacy `create_issue`/`update_issue` names, so they stay green while the live tool slips through. The same end-anchor also drops `sub_issue_write` and any future `*_write` form.
Remedy: do not anchor at the end. Match a backlog noun anywhere AND a mutation verb, e.g. allow when the bare name matches `/(issue|work[_-]?item|backlog[_-]?item|pbi|ticket|story|task)/` AND `/(create|update|edit|write|new|file)/`, keeping the existing comment/reply/cache/search/list/get exclusions. Add `issue_write`, `sub_issue_write` (exclude — no body), and the legacy names to the test matrix.

**2. One-sided verification — the guard is never proven to fire in the real execute path, and the connected happy-path outcome is unverified.**
`src/backend/services/mcp/mcp-orchestrator.screenshot-gating.test.ts:34-86` constructs the orchestrator via `Object.create(MCPOrchestrator.prototype)` and calls the private `normalizeScreenshotMarkdownInArgs` directly. That proves the method's logic but not that `manualLoopAsync` invokes it before `tool.execute` with a connected backlog. The production wiring (`mcp-orchestrator.ts`, ~lines 466-471) is correct on inspection, but no test drives the loop with a body and asserts dedup/caption, and the PR body itself concedes live end-to-end is pending.
Remedy: add the happy-path twin of the bug at `manualLoopAsync` level. The sibling `mcp-orchestrator.false-success.test.ts` already ships a `makeOrchestrator(...)` harness that drives the real loop against scripted tool-call turns; pass `input:{ body: dupeBody }` via its `toolTurn` helper and assert the create-issue execute mock received a body with the image URL exactly once and a `**Figure: ...**` caption. This proves the guard fires in the real path rather than as a unit-callable function.

### Medium

**3. Prompt + backstop are both aimed at a tool name the live server no longer emits, leaving #834 unfixed end-to-end.**
`src/backend/constants/prompts.ts` rule 8 / line 66 still reference `create_issue`, and the deterministic backstop is gated on the predicate in finding 1, which misses `issue_write`. So both halves of the fix target names the production GitHub server doesn't emit. (The prompt prose itself is fine — the load-bearing gap is the predicate plus missing live proof.)
Remedy: align the predicate (finding 1), then run a real shave-to-GitHub flow against `api.githubcopilot.com/mcp` to confirm `normalizeIssueScreenshots` actually executes and the rendered issue shows exactly one captioned screenshot.

**4. Caption-quality regression: a rich top caption is discarded in favour of a weaker synthesised one.**
`src/backend/utils/screenshot-markdown.ts` — `chooseSurvivingOccurrences` always prefers the `### Screenshots` occurrence, and `normalizeIssueScreenshots` only DISCARDS the dropped embed's following Figure caption; it never carries it forward. Repro: top embed `Login page crashes` + caption `**Figure: Login page crashes on submit**`, plus a bare `Login page crashes` under `### Screenshots`. Output caption becomes `**Figure: Login page crashes**` (synthesised from alt text), silently losing "on submit". This is the exact top+Screenshots dedup shape #834 describes, so it is the common case, and it makes the rendered caption worse than what the LLM wrote. The existing test masks it by using a dropped caption equal to the alt text.
Remedy: when a duplicate is dropped, carry its Figure caption forward to the surviving embed if the survivor lacks one (or has only a synthesised/placeholder caption), before calling `buildFigureCaption`.

File: src/backend/utils/screenshot-markdown.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  // Explicit exclusions: these may contain a backlog noun + a create-ish verb but do NOT author
+  // an issue/work-item BODY. Comments/replies carry user-facing text we must not rewrite;
+  // cache/search/list/get/read are read-only. Exclude first so e.g. `add_issue_comment`,
+  // `update_issue_cache`, `create_issue_comment` never slip through.
+  if (/(comment|reply|reaction|cache|search|list|get|read|find|query|delete|close)/.test(name)) {
+    return false;
+  }
+
+  // Allow-list, anchored on the OBJECT of the operation: a backlog create/update tool ends with
+  // the backlog noun it acts on (`create_issue`, `wit_update_work_item`, `jira_create_issue`,
```
Comment (calumjs) [inline src/backend/utils/screenshot-markdown.ts]: The backlog noun is anchored at the END of the bare name, so the live GitHub tool `issue_write` (from the configured api.githubcopilot.com/mcp server) strips to `issue_write`, ends in `write`, fails this anchor, and the screenshot dedupe/caption normalization is silently skipped. `sub_issue_write` and any `*_write` form are dropped too. Match the noun anywhere AND a mutation verb (incl. `write`) instead, and add `issue_write`/`sub_issue_write`/legacy names to the test matrix.

Comment (calumjs) [reply src/backend/utils/screenshot-markdown.ts]: Fixed in ec1c770. `isBacklogItemMutationTool` no longer anchors the backlog noun at the end of the bare name — it now requires a backlog noun **anywhere** AND a body-authoring mutation verb that includes `write`. So the live GitHub `issue_write` tool (from api.githubcopilot.com/mcp) is recognised and the #834 dedupe/caption normalization fires for it. `sub_issue_write` is explicitly excluded (it links sub-issues and has no `body`), as is `issue_read`. Verified empirically: `issue_write`/`create_issue`/`update_issue`/`wit_update_work_item`/`jira_create_issue`/`editJiraIssue` → true; `sub_issue_write`/`add_issue_comment`/`update_issue_cache`/`create_or_update_file`/`create_pull_request`/`list_issues`/`get_issue` → false. Added `issue_write`, `sub_issue_write`, `issue_read` and the legacy names to the unit matrix, plus a manualLoopAsync-level test asserting `issue_write`'s `execute` receives a deduped + captioned body.

File: src/backend/services/mcp/mcp-orchestrator.ts (hunk 1 above)
Comment (calumjs) [reply src/backend/services/mcp/mcp-orchestrator.ts]: Addressed in ec1c770 (building on 4bc5583). The two keyword regexes were already replaced in 4bc5583 by `isBacklogItemMutationTool`; ec1c770 makes it robust against the live GitHub tool name. It now requires a backlog noun anywhere AND a mutation verb, with explicit exclusions FIRST for `*_comment`/`*_reply`/`reaction`/`sub_issue`/`label`/`assign`/`milestone`/`link`/`relation`/`attachment` and all read-only ops (`cache`/`search`/`list`/`get`/`read`/`find`/`query`/`delete`/`close`). The three probed misfires are fixed: `add_issue_comment` → false, `update_issue_cache` → false, and real mutations (`issue_write`, `wit_update_work_item`, `jira_create_issue`) → true. Note the broader `#833` outcome judgement still uses the structured `judgeBacklogOutcome` LLM verdict — this name predicate only gates the deterministic screenshot-markdown rewrite (which body field to normalise), not success/failure. A new manualLoopAsync-level test proves the guard fires in the real execute path and that a comment body passes through unchanged.

File: src/backend/utils/screenshot-markdown.ts (hunk 2 above)
Comment (calumjs) [reply src/backend/utils/screenshot-markdown.ts]: Addressed in ec1c770 (building on 4bc5583). 4bc5583 already replaced keep-first with `chooseSurvivingOccurrences`, which prefers the occurrence inside `### Screenshots` when one exists (so the section is never emptied and the heading isn't left dangling). ec1c770 fixes the remaining caption-quality half you flagged: when the dropped duplicate carried a richer caption than the survivor's alt text, the survivor previously got a weaker alt-synthesised caption. `normalizeIssueScreenshots` now collects the dropped occurrence's Figure caption and carries it forward to a caption-less survivor before falling back to synthesising from alt. Repro from the finding now holds: top `**Figure: Login page crashes on submit**` + bare Screenshots copy → output keeps the Screenshots embed with the full `on submit` caption (added as a regression test), rather than collapsing to `**Figure: Login page crashes**`.

**Review by calumjs — COMMENTED** [review]
## Review: screenshot caption normalization

Verified findings against the PR diff, grouped by severity.

### High

**Caption guard misses an existing caption separated by a blank line — double-caption regression** (`src/backend/utils/screenshot-markdown.ts`, `normalizeIssueScreenshots` hasCaption check)

The "does not double-add a caption when one already exists" guarantee — a headline claim of this PR — breaks when a blank line sits between the embed and an existing `**Figure: ...**` caption. The check only inspects the immediately-next line (`lines[i + 1]`):

```
const next = lines[i + 1];
const hasCaption = next !== undefined && FIGURE_CAPTION_REGEX.test(next);
```

A blank line between the image and the model's own caption (idiomatic, common markdown, and not forbidden by the prompt's "on the next line" wording) defeats detection, so a second alt-derived caption is synthesised directly under the image while the real caption survives below.

Reproduced: input `### Screenshots\nLogin crash\n\n**Figure: Login crash detail**` yields output containing **both** `**Figure: Login crash**` and `**Figure: Login crash detail**` (Figure count = 2). With no blank line the same input correctly yields exactly one caption, isolating the defect to the blank-line gap.

The same `lines[i + 1]`-only adjacency assumption appears in `collectDroppedCaptions` (carry-forward) and in the dedup orphan-cleanup path, so the blind spot is systemic. The existing "does not double-add" test uses an adjacent (zero-blank-line) layout, so the broken branch slips through green.

Remedy: when checking for an existing caption (and when collecting/dropping captions), skip blank/whitespace-only lines and test the next non-empty line against `FIGURE_CAPTION_REGEX` before synthesising. Add a regression test with a blank line between the embed and its `**Figure:**` caption asserting exactly one caption survives.

### Medium

**Dedup and caption-insertion see at most one image per line** (`src/backend/utils/screenshot-markdown.ts`, `chooseSurvivingOccurrences`, `collectDroppedCaptions`, `normalizeIssueScreenshots`)

`IMAGE_LINE_REGEX` is non-global and is applied via `line.match(IMAGE_LINE_REGEX)`, which returns only the first match per line. Multiple inline images on one line is valid markdown, so:
- `Here a and b inline` — only image `a` is seen; only `**Figure: a**` is synthesised; image `b` gets no caption, and the caption is appended as a trailing new line after both images (wrong placement).
- `Crash Crash` (same image duplicated on one line) — output still contains the URL twice; dedup fails entirely. This is the #834 duplicate-screenshot symptom expressed inline.

The "keeps distinct screenshots" test only ever puts one image per line, so this branch is unexercised.

Remedy: iterate all image matches per line with `matchAll` / a global regex, or normalise embeds onto their own lines before processing. Add tests with two distinct images on one line and a duplicate image repeated on one line.

**End-to-end outcome is unverified — one-sided verification** (PR body "Pending live verification"; all three test files)

By the author's own admission live repro/walkthrough is pending and this is "code + build + unit-test only." Every test stubs both the model (`generateTextWithTools`/`generateObject`) and the tool's `execute`. The screenshot-gating-loop test proves the guard's wiring (the deduped body reaches a mocked `execute`) but never exercises the promised outcome: a real shave producing a rendered GitHub issue with exactly one captioned screenshot. Two paths stay unexercised — the `prompts.ts` rule-8 change is an instruction to a live model that is stubbed everywhere, and the real GitHub render path is never driven. This conflicts with the repo's stated "live repro then fix / demonstrate every AC" practice.

Remedy: run the happy-path twin of the bug — a live shave against a connected GitHub backlog — and confirm the rendered issue shows exactly one captioned screenshot end-to-end.

---

Note: the high-severity and the blank-line medium finding are the same root cause; fixing the `hasCaption` lookahead (skipping blank lines) addresses both. The per-line single-image and end-to-end-verification findings are independent.

**Review by tomek-i — APPROVED** [review]
